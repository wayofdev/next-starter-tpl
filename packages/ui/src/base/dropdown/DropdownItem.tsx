import { clsx } from 'clsx'
import Link from 'next/link'
import type { LinkProps } from 'next/link'
import { type ButtonHTMLAttributes, type ReactNode, forwardRef } from 'react'

export enum DropdownItemVariant {
  Button = 'button',
  Link = 'link',
}

export type DropdownItemType = {
  variant?: DropdownItemVariant
  element?: ReactNode
  props?: LinkProps | ButtonHTMLAttributes<HTMLButtonElement>
}

const DropdownItem = forwardRef<HTMLButtonElement, { item: DropdownItemType; isActive?: boolean }>(
  ({ item, isActive = false }, ref) => {
    if (!item.variant) {
      return <>{item.element}</>
    }

    const classNames = clsx(
      'w-full text-left block px-4 py-2 text-sm leading-5 text-gray-700 focus:outline-none transition duration-150 ease-in-out',
      { 'bg-gray-100': isActive }
    )

    const linkProps = (item.props || { href: '#' }) as LinkProps
    const btnProps = (item.props || {}) as ButtonHTMLAttributes<HTMLButtonElement>

    return item.variant === DropdownItemVariant.Link ? (
      <Link {...linkProps} className={classNames}>
        {item.element}
      </Link>
    ) : (
      <button ref={ref} {...btnProps} className={classNames}>
        {item.element}
      </button>
    )
  }
)

DropdownItem.displayName = 'DropdownItem'

export default DropdownItem
