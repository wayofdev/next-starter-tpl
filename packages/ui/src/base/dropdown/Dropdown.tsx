import { Menu, Transition } from '@headlessui/react'
import { Bars3Icon } from '@heroicons/react/24/outline'
import { Fragment } from 'react'
import type { FC, ReactNode } from 'react'
import DropdownItem, { type DropdownItemType } from './DropdownItem'

export enum DropdownAlign {
  Left = 'left',
  Top = 'top',
  Right = 'right',
}

const alignmentClasses: Record<DropdownAlign, string> = {
  [DropdownAlign.Left]: 'origin-top-left left-0',
  [DropdownAlign.Top]: 'origin-top',
  [DropdownAlign.Right]: 'origin-top-right right-0',
}

export type DropdownProps = {
  align?: DropdownAlign
  widthClass?: string
  contentClasses?: string
  trigger?: ReactNode
  items: DropdownItemType[]
}

const Dropdown: FC<DropdownProps> = ({
  align = DropdownAlign.Right,
  widthClass = 'w-48',
  contentClasses = 'py-1 bg-white',
  trigger,
  items,
}) => {
  return (
    <Menu as="div" className="relative">
      {({ open }) => (
        <>
          <Menu.Button as={Fragment}>
            {trigger || (
              <button className="flex items-center text-sm font-medium text-gray-500 transition duration-150 ease-in-out hover:text-gray-700 focus:outline-none">
                <Bars3Icon className="h-6 w-6" aria-hidden="true" />
              </button>
            )}
          </Menu.Button>
          <Transition
            show={open}
            enter="transition ease-out duration-200"
            enterFrom="transform opacity-0 scale-95"
            enterTo="transform opacity-100 scale-100"
            leave="transition ease-in duration-75"
            leaveFrom="transform opacity-100 scale-100"
            leaveTo="transform opacity-0 scale-95"
          >
            <div
              className={`absolute z-50 mt-2 ${widthClass} rounded-md shadow-lg ${alignmentClasses[align]}`}
            >
              <Menu.Items
                className={`rounded-md focus:outline-none ring-1 ring-black ring-opacity-5 ${contentClasses}`}
                static
              >
                {items?.map((item, i) => (
                  <Menu.Item key={i}>
                    {({ active }) => <DropdownItem item={item} isActive={active} />}
                  </Menu.Item>
                ))}
              </Menu.Items>
            </div>
          </Transition>
        </>
      )}
    </Menu>
  )
}

export default Dropdown
