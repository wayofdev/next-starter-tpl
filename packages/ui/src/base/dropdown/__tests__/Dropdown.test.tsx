import renderer from 'react-test-renderer'
import Dropdown, { DropdownAlign } from '../Dropdown'
import { DropdownItemVariant } from '../DropdownItem'

describe('Dropdown', () => {
  const variants = {
    link: { variant: DropdownItemVariant.Link, element: 'Profile', props: { href: '/#Profile' } },
    button: { variant: DropdownItemVariant.Button, element: 'Action' },
    custom: { element: <div className="block w-full bg-gray-800 text-right">Profile</div> },
  }

  Object.values(DropdownAlign).forEach(align => {
    for (const [variantKey, variantValue] of Object.entries(variants)) {
      it(`renders correctly with align=${align}, dropdown item variant=${variantKey}`, () => {
        const tree = renderer
          .create(<Dropdown items={[variantValue]} align={align} widthClass="w-80" />)
          .toJSON()
        expect(tree).toMatchSnapshot()
      })
    }
  })

  it(`renders correctly with custom trigger`, () => {
    const tree = renderer
      .create(
        <Dropdown items={Object.values(variants)} trigger={<button>Custom Trigger</button>} />
      )
      .toJSON()
    expect(tree).toMatchSnapshot()
  })
})
