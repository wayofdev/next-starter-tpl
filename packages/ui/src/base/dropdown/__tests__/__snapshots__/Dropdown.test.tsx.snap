// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dropdown renders correctly with align=left, dropdown item variant=button 1`] = `
<div
  className="relative"
  data-headlessui-state=""
>
  <button
    aria-expanded={false}
    aria-haspopup="menu"
    className="flex items-center text-sm font-medium text-gray-500 transition duration-150 ease-in-out hover:text-gray-700 focus:outline-none"
    data-headlessui-state=""
    id="headlessui-menu-button-:r1:"
    onClick={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
  >
    <svg
      aria-hidden="true"
      className="h-6 w-6"
      fill="none"
      stroke="currentColor"
      strokeWidth={1.5}
      viewBox="0 0 24 24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M3.75 6.75h16.5M3.75 12h16.5m-16.5 5.25h16.5"
        strokeLinecap="round"
        strokeLinejoin="round"
      />
    </svg>
  </button>
</div>
`;

exports[`Dropdown renders correctly with align=left, dropdown item variant=custom 1`] = `
<div
  className="relative"
  data-headlessui-state=""
>
  <button
    aria-expanded={false}
    aria-haspopup="menu"
    className="flex items-center text-sm font-medium text-gray-500 transition duration-150 ease-in-out hover:text-gray-700 focus:outline-none"
    data-headlessui-state=""
    id="headlessui-menu-button-:r2:"
    onClick={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
  >
    <svg
      aria-hidden="true"
      className="h-6 w-6"
      fill="none"
      stroke="currentColor"
      strokeWidth={1.5}
      viewBox="0 0 24 24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M3.75 6.75h16.5M3.75 12h16.5m-16.5 5.25h16.5"
        strokeLinecap="round"
        strokeLinejoin="round"
      />
    </svg>
  </button>
</div>
`;

exports[`Dropdown renders correctly with align=left, dropdown item variant=link 1`] = `
<div
  className="relative"
  data-headlessui-state=""
>
  <button
    aria-expanded={false}
    aria-haspopup="menu"
    className="flex items-center text-sm font-medium text-gray-500 transition duration-150 ease-in-out hover:text-gray-700 focus:outline-none"
    data-headlessui-state=""
    id="headlessui-menu-button-:r0:"
    onClick={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
  >
    <svg
      aria-hidden="true"
      className="h-6 w-6"
      fill="none"
      stroke="currentColor"
      strokeWidth={1.5}
      viewBox="0 0 24 24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M3.75 6.75h16.5M3.75 12h16.5m-16.5 5.25h16.5"
        strokeLinecap="round"
        strokeLinejoin="round"
      />
    </svg>
  </button>
</div>
`;

exports[`Dropdown renders correctly with align=right, dropdown item variant=button 1`] = `
<div
  className="relative"
  data-headlessui-state=""
>
  <button
    aria-expanded={false}
    aria-haspopup="menu"
    className="flex items-center text-sm font-medium text-gray-500 transition duration-150 ease-in-out hover:text-gray-700 focus:outline-none"
    data-headlessui-state=""
    id="headlessui-menu-button-:r7:"
    onClick={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
  >
    <svg
      aria-hidden="true"
      className="h-6 w-6"
      fill="none"
      stroke="currentColor"
      strokeWidth={1.5}
      viewBox="0 0 24 24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M3.75 6.75h16.5M3.75 12h16.5m-16.5 5.25h16.5"
        strokeLinecap="round"
        strokeLinejoin="round"
      />
    </svg>
  </button>
</div>
`;

exports[`Dropdown renders correctly with align=right, dropdown item variant=custom 1`] = `
<div
  className="relative"
  data-headlessui-state=""
>
  <button
    aria-expanded={false}
    aria-haspopup="menu"
    className="flex items-center text-sm font-medium text-gray-500 transition duration-150 ease-in-out hover:text-gray-700 focus:outline-none"
    data-headlessui-state=""
    id="headlessui-menu-button-:r8:"
    onClick={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
  >
    <svg
      aria-hidden="true"
      className="h-6 w-6"
      fill="none"
      stroke="currentColor"
      strokeWidth={1.5}
      viewBox="0 0 24 24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M3.75 6.75h16.5M3.75 12h16.5m-16.5 5.25h16.5"
        strokeLinecap="round"
        strokeLinejoin="round"
      />
    </svg>
  </button>
</div>
`;

exports[`Dropdown renders correctly with align=right, dropdown item variant=link 1`] = `
<div
  className="relative"
  data-headlessui-state=""
>
  <button
    aria-expanded={false}
    aria-haspopup="menu"
    className="flex items-center text-sm font-medium text-gray-500 transition duration-150 ease-in-out hover:text-gray-700 focus:outline-none"
    data-headlessui-state=""
    id="headlessui-menu-button-:r6:"
    onClick={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
  >
    <svg
      aria-hidden="true"
      className="h-6 w-6"
      fill="none"
      stroke="currentColor"
      strokeWidth={1.5}
      viewBox="0 0 24 24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M3.75 6.75h16.5M3.75 12h16.5m-16.5 5.25h16.5"
        strokeLinecap="round"
        strokeLinejoin="round"
      />
    </svg>
  </button>
</div>
`;

exports[`Dropdown renders correctly with align=top, dropdown item variant=button 1`] = `
<div
  className="relative"
  data-headlessui-state=""
>
  <button
    aria-expanded={false}
    aria-haspopup="menu"
    className="flex items-center text-sm font-medium text-gray-500 transition duration-150 ease-in-out hover:text-gray-700 focus:outline-none"
    data-headlessui-state=""
    id="headlessui-menu-button-:r4:"
    onClick={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
  >
    <svg
      aria-hidden="true"
      className="h-6 w-6"
      fill="none"
      stroke="currentColor"
      strokeWidth={1.5}
      viewBox="0 0 24 24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M3.75 6.75h16.5M3.75 12h16.5m-16.5 5.25h16.5"
        strokeLinecap="round"
        strokeLinejoin="round"
      />
    </svg>
  </button>
</div>
`;

exports[`Dropdown renders correctly with align=top, dropdown item variant=custom 1`] = `
<div
  className="relative"
  data-headlessui-state=""
>
  <button
    aria-expanded={false}
    aria-haspopup="menu"
    className="flex items-center text-sm font-medium text-gray-500 transition duration-150 ease-in-out hover:text-gray-700 focus:outline-none"
    data-headlessui-state=""
    id="headlessui-menu-button-:r5:"
    onClick={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
  >
    <svg
      aria-hidden="true"
      className="h-6 w-6"
      fill="none"
      stroke="currentColor"
      strokeWidth={1.5}
      viewBox="0 0 24 24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M3.75 6.75h16.5M3.75 12h16.5m-16.5 5.25h16.5"
        strokeLinecap="round"
        strokeLinejoin="round"
      />
    </svg>
  </button>
</div>
`;

exports[`Dropdown renders correctly with align=top, dropdown item variant=link 1`] = `
<div
  className="relative"
  data-headlessui-state=""
>
  <button
    aria-expanded={false}
    aria-haspopup="menu"
    className="flex items-center text-sm font-medium text-gray-500 transition duration-150 ease-in-out hover:text-gray-700 focus:outline-none"
    data-headlessui-state=""
    id="headlessui-menu-button-:r3:"
    onClick={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
  >
    <svg
      aria-hidden="true"
      className="h-6 w-6"
      fill="none"
      stroke="currentColor"
      strokeWidth={1.5}
      viewBox="0 0 24 24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M3.75 6.75h16.5M3.75 12h16.5m-16.5 5.25h16.5"
        strokeLinecap="round"
        strokeLinejoin="round"
      />
    </svg>
  </button>
</div>
`;

exports[`Dropdown renders correctly with custom trigger 1`] = `
<div
  className="relative"
  data-headlessui-state=""
>
  <button
    aria-expanded={false}
    aria-haspopup="menu"
    data-headlessui-state=""
    id="headlessui-menu-button-:r9:"
    onClick={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
  >
    Custom Trigger
  </button>
</div>
`;
